pkg/operator/controller/callbacks.go

```diff
@@ -253,3 +256,129 @@ func reconcileSetConfigAuthority(args *callbacks.ReconcileCallbackArgs) error {
 
 	return args.Client.Update(context.TODO(), cdi)
 }
+
+func getSpecVersion(version string, crd *extv1.CustomResourceDefinition) *extv1.CustomResourceDefinitionVersion {
+	for _, v := range crd.Spec.Versions {
+		if v.Name == version {
+			return &v
+		}
+	}
+	return nil
+}
+
+func rewriteOldObjects(args *callbacks.ReconcileCallbackArgs, version string, crd *extv1.CustomResourceDefinition) error {
+	args.Logger.Info("Rewriting old objects")
+	kind := crd.Spec.Names.Kind
+	gvk := schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: version,
+		Kind:    kind,
+	}
+	ul := &unstructured.UnstructuredList{}
+	ul.SetGroupVersionKind(gvk)
+	err := args.Client.List(context.TODO(), ul, &client.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, item := range ul.Items {
+		nn := client.ObjectKey{Namespace: item.GetNamespace(), Name: item.GetName()}
+		u := &unstructured.Unstructured{}
+		u.SetGroupVersionKind(item.GetObjectKind().GroupVersionKind())
+		err = args.Client.Get(context.TODO(), nn, u)
+		if err != nil {
+			return err
+		}
+		err = args.Client.Update(context.TODO(), u)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func removeStoredVersion(args *callbacks.ReconcileCallbackArgs, desiredVersion string, crd *extv1.CustomResourceDefinition) error {
+	args.Logger.Info("Removing stored version")
+	crd.Status.StoredVersions = []string{desiredVersion}
+	return args.Client.Status().Update(context.TODO(), crd)
+}
+
+func removeOldVersions(desiredVersion string, crd *extv1.CustomResourceDefinition) *extv1.CustomResourceDefinition {
+	newVersions := make([]extv1.CustomResourceDefinitionVersion, 0)
+	for _, version := range crd.Spec.Versions {
+		if version.Name == desiredVersion {
+			newVersions = append(newVersions, version)
+		}
+	}
+	crd.Spec.Versions = newVersions
+	return crd
+}
+
+// Handle upgrade from clusters that had v1alpha1 as a storage version
+// and remove it from all CRDs managed by us
+func reconcileHandleOldVersion(args *callbacks.ReconcileCallbackArgs) error {
+	if args.State != callbacks.ReconcileStatePostRead {
+		return nil
+	}
+	currentCrd := args.CurrentObject.(*extv1.CustomResourceDefinition)
+	desiredCrd := args.DesiredObject.(*extv1.CustomResourceDefinition)
+	desiredVersion := newestVersion(desiredCrd)
+
+	if olderVersionsExist(desiredVersion, currentCrd) {
+		desiredCrd = restoreOlderVersions(currentCrd, desiredCrd)
+		if !desiredIsStorage(desiredVersion, currentCrd) {
+			// Let kubernetes add it
+			return nil
+		}
+		if err := rewriteOldObjects(args, desiredVersion, currentCrd); err != nil {
+			return err
+		}
+		if err := removeStoredVersion(args, desiredVersion, currentCrd); err != nil {
+			return err
+		}
+	} else {
+		desiredCrd = removeOldVersions(desiredVersion, desiredCrd)
```

Comment:
I think this will be problematic when we support v1 and v1beta1.  I think we can simply remove this else block?

Reply:
This code block exists because the generated CRD still includes v1alpha1 versions, for the purpose of producing release YAMLs that contain v1alpha1 (also: it's several objects because CDI has a CDIConfig field).

Reply:
desiredCrd should not include v1alpha1 in the `else` branch. It is added in the `if` branch.  

Also removeOldVersions is not compatible with serving multiple apiversions.  We will soon want to serve v1 and v1beta1.

Reply:
Any reason that we would want to serve both v1 and v1beta1 and not just migrate to v1 the moment it is introduced using this migration path? it should work just the same.

Reply:
When new versions are introduced, it is common practice to keep the last version around for a while before removing it.  Want existing user yamls to work

Reply:
Here are different things I can do:

1. Remove v1alpha1 from the generated CRD, so this code can be removed. The downside is that someone upgrading a cluster by applying release YAMLs all the way from 1.20.0 will need to remove v1alpha1 as a stored version from the CDI CRD (only this CRD). I provided the command to do so.
2. Make this logic specific to v1alpha1 (hard-coded), so it is not done for v1beta1 once we introduce v1.

Any preference?

Reply:
Is there any way the v1alpha1 version can just be added to the CDI CRD (not DataVolume, etc)?  That way this `else` block can be deleted.  cdi-operator does not do anything with the CDI CRD

Reply:
Done